src/phoenix/trace/llama_index.py

```diff
@@ -0,0 +1,172 @@
+
+For more information on the specification, see
+https://github.com/Arize-ai/open-inference-spec
+"""
+
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Generator, List, Optional, Tuple, TypedDict
+from uuid import UUID, uuid4
+
+from llama_index.callbacks.base_handler import BaseCallbackHandler
+from llama_index.callbacks.schema import CBEvent, CBEventType
+
+from .schemas import Span, SpanKind, SpanStatusCode
+from .tracer import Tracer
+
+
+@dataclass
+class TraceData:
+    id: UUID = field(default_factory=uuid4)
+
+
+class CBEventData(TypedDict, total=False):
+    name: str
+    start_event: CBEvent
+    end_event: CBEvent
+
+
+class OpenInferenceCallbackHandler(BaseCallbackHandler):
+    """Callback handler for storing generation data in OpenInference format.
+    OpenInference is an open standard for capturing and storing AI model
+    inferences. It enables production LLMapp servers to seamlessly integrate
+    with LLM observability solutions such as Arize and Phoenix.
+
+    For more information on the specification, see
+    https://github.com/Arize-ai/open-inference-spec
+    """
+
+    def __init__(self, callback: Optional[Callable[[List[Span]], None]] = None) -> None:
+        super().__init__(event_starts_to_ignore=[], event_ends_to_ignore=[])
+        self._tracer = Tracer(on_append=callback)
+        self._event_id_to_event_data: Dict[str, CBEventData] = defaultdict(lambda: CBEventData())
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        event_id = event_id or str(uuid4())
+        self._event_id_to_event_data[event_id]["name"] = event_type.value
+        self._event_id_to_event_data[event_id]["start_event"] = CBEvent(
+            event_type=event_type,
+            payload=payload,
+            id_=event_id,
+        )
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> None:
+        event_id = event_id or str(uuid4())
+        self._event_id_to_event_data[event_id].setdefault("name", event_type.value)
+        self._event_id_to_event_data[event_id]["end_event"] = CBEvent(
+            event_type=event_type,
+            payload=payload,
+            id_=event_id,
+        )
+
+    def start_trace(self, trace_id: Optional[str] = None) -> None:
+        ...
+
+    def end_trace(
+        self,
+        trace_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        _add_to_tracer(
+            event_id_to_event_data=self._event_id_to_event_data,
+            child_to_parent_trace_id=dict(
+                _generate_child_to_parent_trace_id_pairs(trace_map or {})
```

Comment:
did you mean event_id?
```suggestion
            child_to_parent_event_id=dict(
                _generate_child_to_parent_event_id_pairs(trace_map or {})
```

Reply:
addressed in [5d21bee](https://github.com/Arize-ai/phoenix/pull/1058/commits/5d21bee22395c6d2cf01f00f6d771270fec5fd47)